service-base/src/main/java/org/eclipse/hono/service/tenant/BaseTenantService.java

```diff
@@ -162,16 +162,43 @@ public final void processTenantMessage(final Message<JsonObject> tenantMsg) {
                 remove(tenantId, result -> reply(tenantMsg, result));
                 break;
             default:
-                log.debug("invalid action in request message [{}]", subject);
-                reply(tenantMsg, TenantResult.from(HTTP_BAD_REQUEST));
+                processCustomTenantMessage(tenantMsg, subject);
+                break;
             }
         } catch (final ClassCastException e) {
             log.debug("malformed request message [{}]", e.getMessage());
             reply(tenantMsg, TenantResult.from(HTTP_BAD_REQUEST));
         }
     }
 
-    private void reply(final Message<JsonObject> request, final AsyncResult<TenantResult<JsonObject>> asyncActionResult) {
+    /**
+     * Override the following method to extend TenantService with implementation
+     * specific operations. Consequently, once the operation is completed, a {@link TenantResult} must be
+     * set as reply to this tenantMsg. Use the {@link #reply(Message, AsyncResult)} for sending
+     * the response for the Message. For example, if some concrete action looks like
+     * <pre>{@code
+     *     void doSomeSpecificAction(final Message<JsonObject> tenantMsg, final Handler<AsyncResult<TenantResult>> resultHandler);
+     * }</pre>
+     * then the overriding method would look like
+     *  <pre>{@code
+     *     processCustomTenantMessage(final Message<JsonObject> tenantMsg, String action) {
+     *          if(action.equals("expected-action")) {
+     *            doSomeSpecificAction(tenantMsg, result -> reply(tenantMsg, result));
+     *          } else {
+     *             reply(tenantMsg, TenantResult.from(HttpURLConnection.HTTP_BAD_REQUEST));
+     *         }
+     *       }
+     * }</pre>
+     *
+     * @param tenantMsg target tenant message to be processed.
+     * @param action implementation specific action to be executed.
+     */
+    protected void processCustomTenantMessage(final Message<JsonObject> tenantMsg, String action) {
+        log.debug("invalid action in request message [{}]", action);
+        reply(tenantMsg, TenantResult.from(HTTP_BAD_REQUEST));
+    }
+
+    protected void reply(final Message<JsonObject> request, final AsyncResult<TenantResult<JsonObject>> asyncActionResult) {
```

Comment:
should be declared *final*.
IMHO you need to make the *reply* methods of `BaseRegistrationService` and `BaseCredentialsService` *protected* and *final* as well, so that sub-classes can invoke them, as recommended by your JavaDoc example code.

Reply:
Sorry my bad. I should have seen it coming. I will add it.

Reply:
Sorry, I had it in a different branch and missed actually for this PR.